sold.cc

```diff
@@ -1,985 +1,623 @@
+        for (uintptr_t ptr : bin->init_array()) {
+            init_array_.push_back(ptr + offset);
         }
     }
-
-    void EmitTLS(FILE* fp) {
-        EmitPad(fp, tls_file_offset_);
-        CHECK(ftell(fp) == TLSOffset());
-        for (TLS::Data data : tls_.data) {
-            WriteBuf(fp, data.start, data.size);
+    for (ELFBinary* bin : link_binaries_) {
+        uintptr_t offset = offsets_[bin];
+        for (uintptr_t ptr : bin->fini_array()) {
+            fini_array_.push_back(ptr + offset);
         }
     }
+    LOG(INFO) << "Array numbers: init_array=" << init_array_.size() << " fini_array=" << fini_array_.size();
+}
 
-    void EmitShdr(FILE* fp) {
-        CHECK(ftell(fp) == ShdrOffset());
-        shdr_.EmitShdrs(fp);
-    }
+uintptr_t Sold::RemapTLS(const char* msg, ELFBinary* bin, uintptr_t off) {
+    const Elf_Phdr* tls = bin->tls();
+    CHECK(tls);
+    CHECK(!tls_.data.empty());
+    auto found = tls_.bin_to_index.find(bin);
+    CHECK(found != tls_.bin_to_index.end());
+    const TLS::Data& entry = tls_.data[found->second];
+    if (off < tls->p_filesz) {
+        LOG(INFO) << "TLS data " << msg << " in " << bin->name() << " remapped " << HexString(off) << " => "
+                  << HexString(off + entry.file_offset);
+        off += entry.file_offset;
+    } else {
+        LOG(INFO) << "TLS bss " << msg << " in " << bin->name() << " remapped " << HexString(off) << " => "
+                  << HexString(off + entry.bss_offset);
+        off += entry.bss_offset;
+    }
+    return off;
+}
 
-    void DecideOffsets() {
-        uintptr_t offset = 0x10000000;
-        for (ELFBinary* bin : link_binaries_) {
-            const Range range = bin->GetRange() + offset;
-            CHECK(range.start == offset);
-            offsets_.emplace(bin, range.start);
-            LOG(INFO) << "Assigned: " << bin->soname() << " " << HexString(range.start, 8) << "-" << HexString(range.end, 8);
-            offset = range.end;
-        }
-        tls_offset_ = offset;
-    }
-
-    void CollectTLS() {
-        uintptr_t bss_offset = 0;
-        for (ELFBinary* bin : link_binaries_) {
-            for (Elf_Phdr* phdr : bin->phdrs()) {
-                if (phdr->p_type == PT_TLS) {
-                    uint8_t* start = reinterpret_cast<uint8_t*>(bin->GetPtr(phdr->p_vaddr));
-                    size_t size = phdr->p_filesz;
-                    uintptr_t file_offset = tls_.filesz;
-                    CHECK(tls_.bin_to_index.emplace(bin, tls_.data.size()).second);
-                    tls_.data.push_back({bin, start, size, file_offset, bss_offset});
-                    tls_.memsz += phdr->p_memsz;
-                    tls_.filesz += size;
-                    bss_offset += phdr->p_memsz - size;
-                }
+// Push symbols of bin to symtab.
+// When the same symbol is already in symtab, LoadDynSymtab selects a more
+// concretely defined one.
+// TODO(akawashiro) Raise an error when two same symbols are defined.
```

Comment:
I don't think this is an error. For example, it's legitimate to define `malloc` in the main binary even when it is linked against libc. Adding `LOG(INFO)` would help us debugging the logic, though.

Reply:
I understood. I changed to use `LOG(INFO)`.